### PR TITLE
Colorize vswhere output

### PR DESCRIPTION
**Pull request by heaths**
Print instances to the terminal using colors from Visual Studio Code's Dark+ theme suitable for dark-themed terminals (common). A `-nocolor` switch is added in case it's necessary, but piping stdout to other programs or even using PowerShell accelerators like `[xml]` should work without `-nocolor` since vswhere is not printing directly to the terminal.

### CODE REVIEW

**Comment by heaths** [comment]
Colorizing the output is something I've been wanting to do for a while, and since I did it with the [Go binding's test app](https://github.com/heaths/go-vssetup) I decided to take a crack at it now.

image

image

image

Initially color was enabled by default if stdout was to a terminal and the terminal supported VT100 sequences (which the conhost does in Windows 7 and newer). This works even when output is piped since stdout is no longer to a terminal, and is common in colorized console apps.

Unfortunately, this fails in PowerShell like so:

```powershell
$instances = [xml](vswhere -format xml -utf8)
```

The output is not piped, so the escape sequences are present in the output, which are not valid XML. As such, I added the `-color` switch. It's not idea, but as you can see above makes it much easier to read output, especially in the default formatter.

Maybe someday we bump the major version to 3 and enable colors by default as well as UTF8 output, since that would've been a breaking change as well but probably something we should've done originally when JSON was added.

The colors come from Visual Studio Code's Dark+ theme and should be suitable for the vast majority of terminals, which are generally dark.

**Comment by heaths** [comment]
Windows PowerShell i.e. powershell.exe defaults to a blue background, but these colors are still quite legible:

image

**Comment by heaths** [comment]
Per offline discussion, going back to color by default but with a `-nocolor` escape hatch. I'm not sure what happened such that the `[xml]` accelerator failed when it ran into the first `\033` aka `\1xb` character, but I can't repro it now after doing a completely rebuild of both Debug and Release configurations. I had been running into sporadic incremental build failures with Debug that have always been an issue with this project's .vcxproj references, so that might've been a factor.

The integration tests that run in a Docker container make heavy use of this accelerator so we'll have a better sense for whether it's an issue in an isolated environment.

File: src/vswhere.lib/CommandArgs.h [hunk 1]
```diff
@@ -130,6 +132,12 @@ class CommandArgs
         return m_find;
     }
 
+    const bool get_Color() const noexcept
+
```
Comment (tydunkel) [inline src/vswhere.lib/CommandArgs.h:136]: nit: remove extra newline

File: src/vswhere.lib/JsonFormatter.cpp [hunk 2]
```diff
@@ -26,49 +29,49 @@ wstring JsonFormatter::Escape(_In_ const wstring& value)
...
 
-void JsonFormatter::WriteProperty(_In_ Console& console, _In_ const wstring& name, _In_ const wstring& value)
+void JsonFormatter::WriteProperty(_In_ const wstring& name, _In_ const wstring& value)
 {
-    StartProperty(console, name);
+    StartProperty(name);
 
     auto escaped = Escape(value);
-    console.Write(L"\"%ls\"",escaped.c_str());
+    Console().Write(L"%ls\"%ls\"%ls", Console().Color(ColorValue), escaped.c_str(), Console().ResetColor());
```
Comment (tydunkel) [inline src/vswhere.lib/JsonFormatter.cpp:47]: Might be helpful to create a helper for this since we repeat this set, print, reset pattern throughout

**Review by tydunkel — APPROVED** [review]
LGTM

File: src/vswhere.lib/JsonFormatter.cpp (hunk 2 above)
Comment (heaths) [reply src/vswhere.lib/JsonFormatter.cpp:47]: These are helpers. :) They emit VT100-ish sequences if color is enabled or an empty string.

If you abstract it away much more, you start running into unnecessary compexities. I went through several iterations and this turned out to be the cleanest and leanest.

**Comment by heaths** [comment]
There's one ["problem"](https://github.com/microsoft/terminal/issues/11211) I want to try to sort out first. Doesn't hurt anything, but not fantastic. Seems I need to write and flush to stdout at least once before color attributes take effect. You don't notice it with JSON because of the `[` array start not being colored (like the dark+ theme I mimicked) nor the default text formatter because of the logo (JSON, XML, and value formatters don't use the logo, though value doesn't use color anyway because it's pointless). XML and text with `-nologo` are where you see it.

Perhaps @DHowett has some idea how to resolve the issue. I'm anxious about writing and clearing a single block character because of how that how that might affect output but, then again, that's only when stdout isn't redirected and I'm writing directly to the terminal buffer.

**Review by tydunkel — APPROVED** [review]

**Review by tomsellars70-pixel — CHANGES_REQUESTED** [review]
759e2e54deca5ee124657d2686d0a9e5e27b0f4c

**Pull request merged**